src/process_tree.rs

```diff
@@ -0,0 +1,102 @@
+// SPDX-FileCopyrightText: 2025 blinry <mail@blinry.org>
+//
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+use std::{ops::Deref, path::PathBuf};
+
+use procfs::{process::Process, ProcError, ProcResult};
+
+#[derive(Debug)]
+pub struct CwdProcessTree {
+    tree: ProcessTree,
+    cwd: PathBuf,
+}
+
+#[derive(Debug)]
+pub struct ProcessTree {
+    proc: Process,
+}
+
+impl ProcessTree {
+    pub fn new(pid: u32) -> ProcResult<ProcessTree> {
+        let proc = Process::new(pid as i32)?;
+
+        Ok(ProcessTree { proc })
+    }
+
+    fn children(&self) -> ProcResult<impl IntoIterator<Item = ProcResult<ProcessTree>>> {
+        Ok(self
+            .proc
+            .task_main_thread()?
+            .children()?
+            .into_iter()
+            .map(ProcessTree::new))
+    }
+
+    pub fn into_deepest_leaf(self) -> ProcResult<CwdProcessTree> {
+        fn deepest_leaf(depth: usize, tree: ProcessTree) -> (usize, ProcResult<CwdProcessTree>) {
+            let children = tree.children();
+            let mut max: Option<(usize, ProcResult<CwdProcessTree>)> = None;
+            match children {
+                Ok(children) => {
+                    for child in children {
+                        match child {
+                            Ok(child) => {
+                                let leaf = deepest_leaf(depth + 1, child);
+                                match leaf.1.as_ref() {
+                                    Ok(_) => match max.as_ref() {
+                                        Some(value) => {
+                                            if value.0 < leaf.0 {
+                                                max = Some(leaf);
+                                            }
+                                        }
+                                        None => max = Some(leaf),
+                                    },
+                                    Err(err) => eprintln!("Could not go deeper: {err}"),
+                                }
+                            }
+                            Err(err) => eprintln!("Could not get child: {err}"),
+                        }
+                    }
+                }
+                Err(err) => eprintln!("Could not get children: {err}"),
+            }
```

Comment:
Triangle-shaped code like this is usually a sign that things can be pulled apart a bit more. Maybe instead of collecting the "winning entry" in `max`, we could compute a vector of (depth, process)-tuples? And then, in a second step, take the one with the maximal depth. That would allow us to split this function into multiple ones, because we don't need the `max` context. Do you get what I'm hinting at?

Reply:
I've refactored the method to not use recursion and it's much better now.